hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/ChangeTracker.java

```diff
@@ -135,29 +164,97 @@ public void processResponse(final S3Object object,
         // object was not returned.
         versionMismatches.incrementAndGet();
         throw new RemoteFileChangedException(uri, operation,
-            String.format("%s change "
-                    + CHANGE_REPORTED_BY_S3
+            String.format(CHANGE_REPORTED_BY_S3
                     + " while reading"
                     + " at position %s."
-                    + " Version %s was unavailable",
-                getSource(),
+                    + " %s %s was unavailable",
                 pos,
+                getSource(),
                 getRevisionId()));
       } else {
         throw new PathIOException(uri, "No data returned from GET request");
       }
     }
 
-    final ObjectMetadata metadata = object.getObjectMetadata();
+    processMetadata(object.getObjectMetadata(), operation, pos);
+  }
+
+  /**
+   * Process the response from the server for validation against the
+   * change policy.
+   * @param copyResult result of a copy operation
+   * @throws PathIOException raised on failure
+   * @throws RemoteFileChangedException if the remote file has changed.
+   */
+  public void processResponse(final CopyResult copyResult)
+      throws PathIOException {
+    // ETag (sometimes, depending on encryption and/or multipart) is not the
+    // same on the copied object as the original.  Version Id seems to never
+    // be the same on the copy.  As such, there isn't really anything that
+    // can be verified on the response, except that a revision ID is present
+    // if required.
+    String newRevisionId = policy.getRevisionId(copyResult);
+    LOG.debug("Copy result {}: {}", policy.getSource(), newRevisionId);
+    if (newRevisionId == null && policy.isRequireVersion()) {
+      throw new NoVersionAttributeException(uri, String.format(
+          "Change detection policy requires %s",
+          policy.getSource()));
+    }
+  }
+
+  /**
+   * Process an exception generated against the change policy.
+   * If the exception indicates the file has changed, this method throws
+   * {@code RemoteFileChangedException} with the original exception as the
+   * cause.
+   * @param e the exception
+   * @param operation the operation performed when the exception was
+   * generated (e.g. "copy", "read", "select").
+   * @throws RemoteFileChangedException if the remote file has changed.
+   */
+  public void processException(Exception e, String operation) throws
+      RemoteFileChangedException {
+    if (e instanceof AmazonServiceException) {
+      AmazonServiceException serviceException = (AmazonServiceException) e;
+      if (serviceException.getStatusCode() == 412) {
```

Comment:
Please do not use magic number here - actually, you can use http error codes for this or at least a constant.

Reply:
Don't worry, I'm coming back to this one :)

Btw, I pulled this bit from #606 before similar feedback was provided there.

Reply:
Addressed in https://github.com/apache/hadoop/pull/675/commits/408af6ca1561fdd221da5ddd1a0b50cca829111f

You'll notice I'm referencing the status constant from org.apache.http.HttpStatus, which creates a dependency on org.apache.httpcomponents: httpcore.  That was previously an undeclared (from a maven perspective) dependency that was used only in ITestS3AConfiguration (to reference HttpStatus.SC_MOVED_PERMANENTLY).

I'm curious what you guys think about using that class as the source of HTTP status constants.  I was slightly hesitant since it wasn't already a direct dependency of any of the _source_ files in hadoop-aws.  I'm happy to take another route if you prefer.

S3AUtils has a bunch of magic number status code references that could be replaced by HttpStatus constants if you like this, but that might be something to handle in a separate issue.